src/custom_request_schedule_manager.cc

```diff
@@ -147,4 +150,67 @@ CustomRequestScheduleManager::GetTimestamp(size_t dataset_index) const
   return std::chrono::milliseconds(timestamp_ms);
 }
 
+void
+CustomRequestScheduleManager::PerformWarmup()
+{
+  if (warmup_schedule_.empty()) {
+    throw std::runtime_error("Expected warmup schedule to be non-empty.");
+  }
+
+  const size_t old_max_threads_{max_threads_};
+
+  max_threads_ = std::min(max_threads_, warmup_schedule_.size());
+
+  InitCustomSchedule(warmup_schedule_);
+  WaitForWarmupAndCleanup();
+
+  max_threads_ = old_max_threads_;
```

Comment:
This bit about temporarily changing `max_threads_` was not ideal but the middle-ground implementation that didn't explode a redesign.

In rare cases where there's less than 4 payloads in the warmup, we need to "truncate" `max_threads_` down to that otherwise an infinite loop happens if a thread has no requests to send.

We restore `max_threads_` to its original value in case the user specified something high and the benchmark schedule/dataset is large.